lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php

```diff
@@ -283,19 +293,29 @@ private function removeFromLocalGroup( \User $user, $group ) {
 		return true;
 	}
 
-	public function removeFromGroup( \User $performer, \User $userToChange, $group ) {
-		if ( !$this->canGroupBeRemoved( $performer, $userToChange, $group ) ) {
-			return false;
+	public function removeFromGroup( \User $performer, \User $userToChange, $groups )
+	{
+		$groupList = $groups;
+		if ( !is_array( $groups ) ) {
+			$groupList = [ $groups ];
 		}
 
-		try {
-			if ( in_array( $group, $this->permissionsConfiguration->getGlobalGroups() ) ) {
-				$result = $this->removeFromGlobalGroup( $userToChange, $group );
-			} else {
-				$result = $this->removeFromLocalGroup( $userToChange, $group );
+		//First check if we can remove all groups (if we remove in parallel to checking, then the check may not be valid)
+		foreach ( $groupList as $group ) {
+			if ( !$this->canGroupBeRemoved( $performer, $userToChange, $group ) ) {
+				return false;
 			}
 		}
-		finally {
+		$result = true;
+		try {
+			foreach ( $groupList as $group ) {
+				if ( in_array( $group, $this->permissionsConfiguration->getGlobalGroups() ) ) {
+					$result = $this->removeFromGlobalGroup( $userToChange, $group ) && $result;
+				} else {
+					$result = $this->removeFromLocalGroup( $userToChange, $group ) && $result;
+				}
+			}
+		} finally {
```

Comment:
That's a good question - the only problem (at least for now, until we migrate to external service) that I can think of here would be a problem with DB - I'm not sure if there's something more we should do here in such case? I think that in such case the exception should simply propagate higher - that's basically also the way it was working before. I added the finally statement to make sure we don't have some stale data cached locally. Is there something more that you'd suggest?